callback_test.go

```diff
@@ -0,0 +1,50 @@
+//go:build manual_test
+
+// This test requires manually bulding libcbtest.so as a prerequisite so it is disabled by default.
+//     make -C libcbtest
+//     go test -tags=manual_test -v -run ^TestCallGoFromSharedLib$ .
+
+package purego_test
+
+import (
+	"sync"
+	"testing"
+	"unsafe"
+
+	"github.com/ebitengine/purego"
+)
+
+// TestCallGoFromSharedLib is a test that checks for stack corruption on darwin/arm64
+// when C calls Go code from a non-Go thread in a dynamically loaded share library.
+func TestCallGoFromSharedLib(t *testing.T) {
+	const libname = "libcbtest/libcbtest.so"
+	lib, err := purego.Dlopen(libname, purego.RTLD_NOW|purego.RTLD_GLOBAL)
+	if err != nil {
+		t.Fatalf("Dlopen(%q) faled: %v", libname, err)
+	}
+
+	var initLib func()
+	purego.RegisterLibFunc(&initLib, lib, "initLib")
+	initLib()
+
+	var printString func(p uintptr, s string)
+	purego.RegisterLibFunc(&printString, lib, "printString")
+
+	var wg sync.WaitGroup
+
+	goFunc := func(cstr *byte, n int) int {
+		a := (*[1<<30 - 1]byte)(unsafe.Pointer(cstr))[:n]
```

Comment:
This could be `string(unsafe.Slice(cstr, n))`

Reply:
Done. Since `unsafe.String` requires 1.20, I falsely assumed the same for `unsafe.Slice`.